src/merkle/store/mod.rs

```diff
@@ -280,6 +277,37 @@ impl<T: KvMap<RpoDigest, StoreNode>> MerkleStore<T> {
         })
     }
 
+    /// Iterator over the non-empty leaves of the Merkle tree associated with the specified `root`
+    /// and `max_depth`.
+    pub fn non_empty_leaves(
+        &self,
+        root: RpoDigest,
+        max_depth: u8,
+    ) -> impl Iterator<Item = (NodeIndex, RpoDigest)> + '_ {
+        let empty_roots = EmptySubtreeRoots::empty_hashes(max_depth);
+        let mut stack = Vec::new();
+        stack.push((NodeIndex::new_unchecked(0, 0), root));
+
+        core::iter::from_fn(move || {
+            while let Some((index, node_hash)) = stack.pop() {
+                if index.depth() == max_depth {
+                    return Some((index, node_hash));
+                }
+
+                if let Some(node) = self.nodes.get(&node_hash) {
+                    if !empty_roots.contains(&node.left) {
+                        stack.push((index.left_child(), node.left));
+                    }
+                    if !empty_roots.contains(&node.right) {
+                        stack.push((index.right_child(), node.right));
+                    }
+                }
+            }
+
+            None
+        })
```

Comment:
I think we may need to refactor this to support the tiered SMT.  At the minute the termination condition for identifying leaves is:

```rust
                if index.depth() == max_depth {
                    return Some((index, node_hash));
                }
```

However for tiered SMT we can have leaves at `depth < max_depth`.  I think we have a two ways of addressing this.  Assume that if for some `depth < max_depth` `self.nodes.get(&node_hash) == None` then we have found a leaf.  We would modify the iterator code as follows:

```rust
        core::iter::from_fn(move || {
            while let Some((index, node_hash)) = stack.pop() {
                if index.depth() == max_depth {
                    return Some((index, node_hash));
                }

                if let Some(node) = self.nodes.get(&node_hash) {
                    if !empty_roots.contains(&node.left) {
                        stack.push((index.left_child(), node.left));
                    }
                    if !empty_roots.contains(&node.right) {
                        stack.push((index.right_child(), node.right));
                    }
                } else {                     <---- add this condition
                    return Some((index, node_hash));
                }
            }

            None
        })
```

However, this uses implicit assumption that if a node hash isn't in the store then it must be a node. An more rigorous approach would be to modify the store as follows.  First we modify `StoreNode` as follows:

```rust
pub enum StoreNode {
    InnerNode { left: RpoDigest, right: RpoDigest },
    LeafNode { value: RpoDigest },
}
```

We could then modify the `non_empty_leaves` function in the following way (we would also need to introduce a condition to error if `max_depth` is reached without finding a leaf):
```rust
    pub fn non_empty_leaves(
        &self,
        root: RpoDigest,
        max_depth: u8,
    ) -> impl Iterator<Item = (NodeIndex, RpoDigest)> + '_ {
        let empty_roots = EmptySubtreeRoots::empty_hashes(max_depth);
        let mut stack = Vec::new();
        stack.push((NodeIndex::new_unchecked(0, 0), root));

        core::iter::from_fn(move || {
            while let Some((index, node_hash)) = stack.pop() {
                if let Some(node) = self.nodes.get(&node_hash) {
                    match node {
                        StoreNode::InnerNode { left, right } => {
                            if !empty_roots.contains(left) {
                                stack.push((index.left(), *left));
                            }
                            if !empty_roots.contains(right) {
                                stack.push((index.right(), *right));
                            }
                        }
                        StoreNode::LeafNode { value } => {
                            return Some((index, *value));
                        }
                    }
                } else {
                    return Err(MerkleError::NodeNotInStore(node_hash, index));
                }
            }

            None
        })
    }
```

We would also need to modify the way in which we embed child trees in parent trees.  Instead of directly placing the root of the child tree as a leaf in the parent tree, we would hash the root and then insert the hashed root as a leaf in the parent tree.  This is a pretty big refactor but I think it's worth considering as we currently lack node context in the `MerkleStore`.

Reply:
I was thinking more of the first approach you outlined and I think it should work for TSMT use case. Leaves there are hashed in different domains, and thus, will never have child nodes in the `MerkleStore` itself (assuming the hash function is not broken).

But the other approach is worth considering - let's create a dedicated issue for this?

Reply:
done #172